File: 11-observability/python-sdk/basic-tracing.py
"""
Cost Katana OpenTelemetry: Basic Distributed Tracing (Python)

Track AI requests across services with distributed tracing.

Run: python 11-observability/python-sdk/basic-tracing.py
"""

import os
import uuid
import requests

API_BASE = "https://cost-katana-backend.store/api"
GATEWAY_URL = f"{API_BASE}/gateway/v1/chat/completions"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def basic_traced_request():
    """Make a traced AI request"""
    
    session_id = f"session_{uuid.uuid4().hex[:8]}"
    
    print(f"1️⃣ Making traced request...")
    print(f"   Session ID: {session_id}")
    
    response = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [
                {"role": "user", "content": "Explain distributed tracing"}
            ]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json",
            "X-Session-Id": session_id,
            "X-Service-Name": "my-python-app"
        }
    )
    
    response.raise_for_status()
    
    trace_id = response.headers.get("X-Trace-Id")
    duration = response.headers.get("X-Request-Duration")
    cost = response.headers.get("X-Cost")
    
    print(f"✅ Request traced:")
    print(f"   Trace ID: {trace_id}")
    print(f"   Duration: {duration}")
    print(f"   Cost: ${cost}")
    
    return {
        "session_id": session_id,
        "trace_id": trace_id,
        "cost": cost
    }

def multi_request_session():
    """Track multiple requests in a session"""
    
    session_id = f"multi_session_{uuid.uuid4().hex[:8]}"
    
    print(f"\n2️⃣ Multi-request session...")
    print(f"   Session ID: {session_id}")
    
    traces = []
    
    # Request 1
    print("\n   Request 1: Extract entities...")
    r1 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-4",
            "messages": [{"role": "user", "content": "Extract entities..."}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Session-Id": session_id,
            "X-Span-Name": "entity_extraction"
        }
    )
    r1.raise_for_status()
    traces.append({
        "operation": "entity_extraction",
        "trace_id": r1.headers.get("X-Trace-Id"),
        "cost": r1.headers.get("X-Cost")
    })
    
    # Request 2 (child of Request 1)
    print("   Request 2: Classify sentiment...")
    r2 = requests.post(
        GATEWAY_URL,
        json={
            "model": "gpt-3.5-turbo",
            "messages": [{"role": "user", "content": "Classify sentiment..."}]
        },
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "X-Session-Id": session_id,
            "X-Parent-Trace-Id": r1.headers.get("X-Trace-Id"),
            "X-Span-Name": "sentiment_classification"
        }
    )
    r2.raise_for_status()
    traces.append({
        "operation": "sentiment_classification",
        "trace_id": r2.headers.get("X-Trace-Id"),
        "parent_id": r1.headers.get("X-Trace-Id"),
        "cost": r2.headers.get("X-Cost")
    })
    
    total_cost = sum(float(t["cost"] or 0) for t in traces)
    
    print(f"\n✅ Session complete:")
    print(f"   Total traces: {len(traces)}")
    print(f"   Total cost: ${total_cost:.4f}")
    
    return traces

def view_session_trace(session_id):
    """Retrieve and display session trace"""
    
    print(f"\n3️⃣ Viewing session trace: {session_id}")
    
    response = requests.get(
        f"{API_BASE}/v1/sessions/{session_id}",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    session = response.json()
    
    print(f"✅ Session trace:")
    print(f"   Total Traces: {len(session.get('traces', []))}")
    print(f"   Total Cost: ${session.get('totalCost', 0)}")
    print(f"   Total Duration: {session.get('totalDuration', 0)}ms")
    
    if session.get('traces'):
        print("\n   Trace Hierarchy:")
        for trace in session['traces']:
            indent = "   └─" if trace.get('parentId') else "   ├─"
            print(f"{indent} {trace.get('operation', 'request')}")
            print(f"      Trace ID: {trace.get('traceId')}")
            print(f"      Duration: {trace.get('duration')}ms")
            print(f"      Cost: ${trace.get('cost')}")
    
    return session

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    print("🥷 Cost Katana OpenTelemetry: Basic Tracing (Python)\n")
    
    try:
        # Run examples
        result = basic_traced_request()
        multi_request_session()
        view_session_trace(result["session_id"])
        
        print("\n✅ All tracing examples complete!")
        print("\n💡 Benefits:")
        print("   • Track requests across services")
        print("   • Monitor AI performance")
        print("   • Debug with full context")
        print("   • Attribute costs accurately")
        
    except requests.exceptions.HTTPError as e:
        print(f"❌ Error: {e.response.json() if e.response else e}")

if __name__ == "__main__":
    main()

File: 11-observability/npm-package/basic-tracing.ts
/**
 * Cost Katana OpenTelemetry: Basic Distributed Tracing
 * 
 * Enable distributed tracing across your AI applications to track
 * requests, monitor performance, and attribute costs.
 * 
 * Setup:
 * 1. npm install cost-katana @opentelemetry/api
 * 2. Set COST_KATANA_API_KEY
 * 3. Run: npx ts-node 11-observability/npm-package/basic-tracing.ts
 */

import CostKatana from 'cost-katana';
import { trace, context, SpanStatusCode } from '@opentelemetry/api';
import { v4 as uuidv4 } from 'uuid';

const client = new CostKatana({
  apiKey: process.env.COST_KATANA_API_KEY!
});

/**
 * Example 1: Basic traced AI request
 */
async function basicTracedRequest() {
  const sessionId = uuidv4();
  
  console.log('1️⃣ Making traced AI request...');
  console.log(`   Session ID: ${sessionId}`);

  try {
    const response = await client.chat.completions.create({
      model: 'gpt-4',
      messages: [
        { role: 'user', content: 'Explain distributed tracing in AI systems' }
      ]
    }, {
      // Trace context headers
      headers: {
        'X-Session-Id': sessionId,
        'X-Service-Name': 'my-ai-app'
      }
    });

    console.log('✅ Response received with trace context:');
    console.log(`   Trace ID: ${response.headers?.['x-trace-id']}`);
    console.log(`   Duration: ${response.headers?.['x-request-duration']}`);
    console.log(`   Cost: $${response.headers?.['x-cost']}`);
    console.log(`   Tokens: ${response.usage?.total_tokens}`);

    return {
      sessionId,
      traceId: response.headers?.['x-trace-id'],
      cost: response.headers?.['x-cost']
    };
  } catch (error) {
    console.error('❌ Request failed:', error);
    throw error;
  }
}

/**
 * Example 2: Multi-request session tracking
 */
async function multiRequestSession() {
  const sessionId = `session_${uuidv4()}`;
  
  console.log('\n2️⃣ Starting multi-request session...');
  console.log(`   Session ID: ${sessionId}`);

  const traces = [];

  try {
    // Request 1: Extract entities
    console.log('\n   Request 1: Extract entities...');
    const response1 = await client.chat.completions.create({
      model: 'gpt-4',
      messages: [
        { role: 'user', content: 'Extract entities from: John works at Acme Corp in New York' }
      ]
    }, {
      headers: {
        'X-Session-Id': sessionId,
        'X-Span-Name': 'entity_extraction'
      }
    });
    traces.push({
      operation: 'entity_extraction',
      traceId: response1.headers?.['x-trace-id'],
      duration: response1.headers?.['x-request-duration'],
      cost: response1.headers?.['x-cost']
    });

    // Request 2: Classify sentiment (child of session)
    console.log('   Request 2: Classify sentiment...');
    const response2 = await client.chat.completions.create({
      model: 'gpt-3.5-turbo',
      messages: [
        { role: 'user', content: 'Classify sentiment: This product is amazing!' }
      ]
    }, {
      headers: {
        'X-Session-Id': sessionId,
        'X-Parent-Trace-Id': response1.headers?.['x-trace-id'],
        'X-Span-Name': 'sentiment_classification'
      }
    });
    traces.push({
      operation: 'sentiment_classification',
      traceId: response2.headers?.['x-trace-id'],
      parentId: response1.headers?.['x-trace-id'],
      duration: response2.headers?.['x-request-duration'],
      cost: response2.headers?.['x-cost']
    });

    console.log('\n✅ Session complete:');
    console.log(`   Total traces: ${traces.length}`);
    console.log(`   Total cost: $${traces.reduce((sum, t) => sum + parseFloat(t.cost || '0'), 0).toFixed(4)}`);
    
    // View session details
    const session = await viewSessionTrace(sessionId);
    return session;

  } catch (error) {
    console.error('❌ Session failed:', error);
    throw error;
  }
}

/**
 * Example 3: W3C Trace Context propagation
 */
async function w3cTraceContext() {
  console.log('\n3️⃣ Using W3C Trace Context standard...');

  // Generate W3C compliant trace context
  const traceId = generateTraceId();
  const spanId = generateSpanId();
  const traceparent = `00-${traceId}-${spanId}-01`;

  console.log(`   traceparent: ${traceparent}`);

  try {
    const response = await client.chat.completions.create({
      model: 'gpt-4',
      messages: [
        { role: 'user', content: 'What is W3C Trace Context?' }
      ]
    }, {
      headers: {
        'traceparent': traceparent,
        'tracestate': 'costkatana=tracked,vendor=myapp'
      }
    });

    console.log('✅ Request traced with W3C standard');
    console.log(`   Cost Katana Trace ID: ${response.headers?.['x-trace-id']}`);
    console.log(`   W3C Trace ID: ${traceId}`);

    return response;
  } catch (error) {
    console.error('❌ Failed:', error);
    throw error;
  }
}

/**
 * Example 4: View session traces
 */
async function viewSessionTrace(sessionId: string) {
  console.log(`\n4️⃣ Retrieving session trace: ${sessionId}`);

  try {
    const response = await fetch(
      `https://cost-katana-backend.store/api/v1/sessions/${sessionId}`,
      {
        headers: {
          'Authorization': `Bearer ${process.env.COST_KATANA_API_KEY}`
        }
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to get session: ${response.status}`);
    }

    const session = await response.json();

    console.log('✅ Session trace retrieved:');
    console.log(`   Session ID: ${session.sessionId}`);
    console.log(`   Total Traces: ${session.traces?.length || 0}`);
    console.log(`   Total Cost: $${session.totalCost || 0}`);
    console.log(`   Total Duration: ${session.totalDuration || 0}ms`);

    if (session.traces) {
      console.log('\n   Trace Hierarchy:');
      session.traces.forEach((trace: any, index: number) => {
        const indent = trace.parentId ? '   └─' : '   ├─';
        console.log(`${indent} ${trace.operation || 'request'}`);
        console.log(`      Trace ID: ${trace.traceId}`);
        console.log(`      Duration: ${trace.duration}ms`);
        console.log(`      Cost: $${trace.cost}`);
      });
    }

    return session;
  } catch (error: any) {
    console.error('❌ Failed to retrieve session:', error.message);
    throw error;
  }
}

/**
 * Example 5: Error tracking with traces
 */
async function tracedErrorHandling() {
  const sessionId = `error_session_${uuidv4()}`;
  
  console.log('\n5️⃣ Demonstrating error tracking...');

  try {
    // This will fail - invalid model
    await client.chat.completions.create({
      model: 'invalid-model-name',
      messages: [{ role: 'user', content: 'Test' }]
    }, {
      headers: {
        'X-Session-Id': sessionId,
        'X-Span-Name': 'error_test'
      }
    });
  } catch (error: any) {
    console.log('✅ Error captured in trace:');
    console.log(`   Session ID: ${sessionId}`);
    console.log(`   Error: ${error.message}`);
    console.log(`   Error is automatically recorded in trace span`);
    
    // View the failed trace
    await viewSessionTrace(sessionId);
  }
}

/**
 * Utility: Generate W3C trace ID (128-bit hex)
 */
function generateTraceId(): string {
  return Array.from({ length: 32 }, () => 
    Math.floor(Math.random() * 16).toString(16)
  ).join('');
}

/**
 * Utility: Generate W3C span ID (64-bit hex)
 */
function generateSpanId(): string {
  return Array.from({ length: 16 }, () => 
    Math.floor(Math.random() * 16).toString(16)
  ).join('');
}

/**
 * Main example runner
 */
async function main() {
  if (!process.env.COST_KATANA_API_KEY) {
    console.error('❌ COST_KATANA_API_KEY environment variable required');
    process.exit(1);
  }

  console.log('🥷 Cost Katana OpenTelemetry: Basic Tracing\n');

  try {
    // Run examples
    await basicTracedRequest();
    await multiRequestSession();
    await w3cTraceContext();
    await tracedErrorHandling();

    console.log('\n✅ All tracing examples complete!');
    console.log('\n💡 Next steps:');
    console.log('   - View traces in Cost Katana dashboard');
    console.log('   - Export to Jaeger/Zipkin for visualization');
    console.log('   - Set up alerts on trace metrics');
    console.log('   - Integrate with your APM solution');
    console.log('\n📊 Benefits:');
    console.log('   - Track request flows across services');
    console.log('   - Monitor AI performance in real-time');
    console.log('   - Debug issues with full context');
    console.log('   - Attribute costs accurately');

  } catch (error) {
    console.error('\n❌ Example failed:', error);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

export {
  basicTracedRequest,
  multiRequestSession,
  w3cTraceContext,
  viewSessionTrace,
  tracedErrorHandling
};

